d3ploy/no_inst.py

```diff
@@ -152,23 +166,68 @@ def tock(self):
                 continue
             diff, supply, demand = self.calc_diff(commod, time-1)
             if  diff < 0:
-                proto = random.choice(self.commod_to_fac[commod])
-                ## This is still not correct. If no facilities are present at the start of the
-                ## simulation prod_rate will still return zero. More complex fix is required.
-                if proto in self.fac_supply[commod]:
-                    prod_rate = self.fac_supply[commod][proto]
-                else:
-                    print("No facility production rate available for " + proto)
-                number = np.ceil(-1*diff/prod_rate)
-                for i in range(int(number)):
-                    self.context.schedule_build(self, proto)
-                    i += 1
+                deploy_dict = self.deploy_solver(commod, diff)
+                for proto, num in deploy_dict.items():
+                    for i in range(num):                        
+                        self.context.schedule_build(self, proto)
             if self.record:
                 out_text = "Time " + str(time) + " Deployed " + str(len(self.children))
                 out_text += " supply " + str(self.commodity_supply[commod][time-1])
                 out_text += " demand " + str(self.commodity_demand[commod][time-1]) + "\n"
                 with open(commod +".txt", 'a') as f:
                     f.write(out_text)
+    
+        
+    def deploy_solver(self, commod, diff):
+        """ This function optimizes prototypes to deploy to minimize over
+            deployment of prototypes.
+
+        Paramters:
+        ----------
+        commod: str
+            commodity to deploy the prototypes for
+        diff: float
+            lack in supply
+        
+        Returns:
+        --------
+        deploy_dict: dict
+            key: prototype name
+            value: # to deploy
+        """
+        proto_commod = self.commodities[commod]
+        min_cap = min(proto_commod.values())
+        if diff < min_cap:
+            return {}
+
+        key_list = self.get_asc_key_list(proto_commod)
+
+        after = diff
```

Comment:
no crucial, but remaining, or remainder might be a better variable name than after